### PR TITLE
63 refactor colorshadowedcard

### PR DESCRIPTION
**Pull request by Kszesiek**
What happened:
- Created new widget ImageWithOverlay for the purpose of overlaying semi-opaque text over image (suggested to use with ColorShadowedCard, SwipeableCard and everywhere else where applicable)
- Used the new widget in SwipeableCard & done some cleaning there
- Used the new widget in articles page
- Major cleanup on the homescreen - exported all the remaining mockup widgets to separate files
- Refactored mockup widgets from the homescreen to use ColorShadowedCard, SwipeableCard and ImageWithOverlay so they can be further used in the current form (after creating fetching logic)
- Got rid of ColorShadowedCard2 for good
- Ironically ColorShadowedCard didn't change at all lol

How things look like atm:
obraz

### CODE REVIEW

File: mobile/lib/components/utility/image_with_overlay.dart [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+/// Widget representing an image with overlaid text on the bottom of the image.
+class ImageWithOverlay extends StatelessWidget {
+  const ImageWithOverlay({
+    super.key,
+    this.child,
+    required this.thumbnailPath,
+    this.imageConstructor = Image.network,
+    this.titleOverlay,
+    required this.isLoading,
+    this.titleOverlayPadding = const EdgeInsets.all(4),
```
Comment (julianuziemblo) [inline mobile/lib/components/utility/image_with_overlay.dart:14]: just for clarity, the constructor parameters should be in order, e.g.: super first, then required, then with no value, and last with default value

File: mobile/lib/components/utility/image_with_overlay.dart [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+  /// Child widget to display centered above the image, BUT ABOVE TITLE OVERLAY.
+  final Widget? child;
+
+  /// Network or local path to the image.
+  final String thumbnailPath;
+
+  /// Function used to display the image (used mainly to be able to switch
+  /// between network and local images). Defaults to network image.
+  final Image Function(String, {BoxFit fit}) imageConstructor;
```
Comment (julianuziemblo) [inline mobile/lib/components/utility/image_with_overlay.dart:25]: this pattern in flutter is usually called something more like "imageBuilder"

File: mobile/lib/components/utility/image_with_overlay.dart [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        // if (child != null)
+        //   Positioned.fill(
+        //     child: child!,
+        //   ),
+        Positioned.fill(
+          child: Column(
+            crossAxisAlignment: CrossAxisAlignment.stretch,
+            children: [
+              Expanded(
+                child: child ?? Container(),
```
Comment (julianuziemblo) [inline mobile/lib/components/utility/image_with_overlay.dart:63]: const SizedBox.shrink() would be suitable more for a dummy widget that will not be displayed

File: mobile/lib/components/utility/image_with_overlay.dart [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                : imageConstructor(
+                    thumbnailPath,
+                    fit: BoxFit.cover,
+                  ),
+          ),
+        ),
+        // if (child != null)
+        //   Positioned.fill(
+        //     child: child!,
+        //   ),
```
Comment (julianuziemblo) [inline mobile/lib/components/utility/image_with_overlay.dart:57]: This should be deleted I suppose

File: mobile/lib/components/utility/image_with_overlay.dart [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            color: context.colors.backgroundDarkSecondary,
+            child: isLoading
+                ? Image.asset(
+                    'assets/defaultMedia.png',
+                    fit: BoxFit.cover,
+                  )
+                : imageConstructor(
+                    thumbnailPath,
+                    fit: BoxFit.cover,
+                  ),
```
Comment (julianuziemblo) [inline mobile/lib/components/utility/image_with_overlay.dart:51]: We would have to ask RA on this, but I think there are 3 states at least - loading, success and fail. I think you should let the user handle all 3 (because it is really easy thanks to `Image.<something>()`'s optional builder methods).
tl;dr: would recommend to just leave it as an image with no state and let the user handle those states.

File: mobile/lib/components/teraz_gramy/teraz_gramy_widget.dart [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  const TerazGramyWidget({
+    super.key,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    return ColorShadowedCard(
+      shadowColor: context.colors.highlightRed,
+      child: ImageWithOverlay(
+        isLoading: false,
```
Comment (julianuziemblo) [inline mobile/lib/components/teraz_gramy/teraz_gramy_widget.dart:21]: this illustrates my point from `lib/components/utility/image_with_overlay.dart`'s comments: it's kinda awkward to just say "isLoading: false".

File: mobile/lib/components/newest_recording/newest_recording_widget.dart [hunk 7]
```diff
@@ -0,0 +1,45 @@
+import 'package:flutter/cupertino.dart';
+import 'package:radioaktywne/components/utility/swipeable_card.dart';
+import 'package:radioaktywne/extensions/build_context.dart';
+
+/// Widget representing three most recent recordings.
```
Comment (julianuziemblo) [inline mobile/lib/components/newest_recording/newest_recording_widget.dart:5]: add a TODO comment explaining this is a dummy widget - it will make it easier to make the app test-ready

File: mobile/lib/main.dart [hunk 8]
```diff
@@ -59,10 +60,11 @@ class MainPage extends StatelessWidget {
       color: context.colors.backgroundLight,
       width: double.infinity,
       height: double.infinity,
-      child: SingleChildScrollView(
+      child: const SingleChildScrollView(
         child: Padding(
-          padding: const EdgeInsets.only(
+          padding: EdgeInsets.only(
             bottom: RaPageConstraints.radioPlayerHeight,
+            top: 8,
```
Comment (julianuziemblo) [inline mobile/lib/main.dart:67]: add constant for this in `lib/resources/ra_page_constraints.dart`

**Review by julianuziemblo — CHANGES_REQUESTED** [review]
Works good, but some things are missing compared to previous versions. However, those things weren't agreed on with RA so we will have to ask.

Also a lot of nitpicks. Sorry.

File: mobile/lib/pages/article_selection_page.dart [hunk 9]
```diff
@@ -48,11 +49,13 @@ class ArticleSelectionPage extends HookWidget {
       ), // Helps with the player not covering the last article
       child: GridView.builder(
         controller: hooks.scrollController,
-        padding: const EdgeInsets.all(20),
+        padding: RaPageConstraints.outerWidgetPagePadding.copyWith(
+          top: RaPageConstraints.pagePadding,
+        ),
```
Comment (julianuziemblo) [inline mobile/lib/pages/article_selection_page.dart:54]: One last thing: add a bottom padding here (possibly `RaPageConstraints.radioPlayerPadding`), because it clashes with the radio player when scrolled to the bottom

File: mobile/lib/components/utility/image_with_overlay.dart (hunk 5 above)
Comment (Kszesiek) [reply mobile/lib/components/utility/image_with_overlay.dart:51]: I managed to rewrite it with errorBuilder and loadingBuilder in mind, have a look and see if that's what it's supposed to look like.

File: mobile/lib/main.dart (hunk 8 above)
Comment (Kszesiek) [reply mobile/lib/main.dart:67]: Found another suitable solution actually

**Comment by Kszesiek** [comment]
> Also a lot of nitpicks. Sorry.

That's probably the PR where nitpicks are expected, if not even encouraged 😂

**Review by julianuziemblo — APPROVED** [review]
Good 2 go!
